### PR TITLE
chore(dev): Improve low Code and Refactor Functions

### PR DESCRIPTION
**Pull request by yezz123**
Hello @adriangb It's been a long time we talk about the projects, I just find the right time to do some improvement in code may be related to styling and improving code to use the last Pythonic ways, that's why!

I guess you will find most of the refactor functions are:
- Simplifies boolean if expressions by removing unnecessary explicit references to True or False states.
- Replaces conditional assignment to a variable with an if expression

And other ones, also sorry for this Big Pull request 😅

### CODE REVIEW

File: xpresso/applications.py [hunk 1]
```diff
@@ -159,7 +159,7 @@ async def __call__(
         send: starlette.types.Send,
     ) -> None:
         scope_type = scope["type"]
-        if scope_type == "http" or scope_type == "websocket":
+        if scope_type in ["http", "websocket"]:
```
Comment (adriangb) [inline xpresso/applications.py:162]: This one is very on purpose: it is going to be much faster to compare to two items (and if it is `"http"`, which is going to be most of the request, continue after 1 comparison) than it is to always create a 2 element list (allocated on the heap, pretty slow) and then do an `in` (which probably converts the list to an utterable under the hood and stuff, even more overhead)

File: xpresso/_utils/media_type_validator.py [hunk 2]
```diff
@@ -24,10 +24,7 @@ def matches(self, media_type: typing.Optional[str]) -> bool:
         if media_type is None:
             return False
         media_type = next(iter(media_type.split(";"))).lower()
-        for accepted in self.accepted:
-            if accepted.match(media_type):
-                return True
-        return False
+        return any(accepted.match(media_type) for accepted in self.accepted)
```
Comment (adriangb) [inline xpresso/_utils/media_type_validator.py:27]: I don't care much either way. I tend to prefer the more verbose version that doesn't use language specific features (remember that `any` is a builtin and that you're actually creating a generator in there), but I'd be okay replacing it.

File: xpresso/applications.py [hunk 3]
```diff
@@ -249,8 +249,9 @@ def _get_doc_routes(
             async def openapi(req: Request) -> JSONResponse:
                 if self._openapi is None:
                     self._openapi = self.get_openapi()
-                res = JSONResponse(self._openapi.dict(exclude_none=True, by_alias=True))
-                return res
+                return JSONResponse(
+                    self._openapi.dict(exclude_none=True, by_alias=True)
+                )
```
Comment (adriangb) [inline xpresso/applications.py:254]: 👍

File: xpresso/applications.py [hunk 4]
```diff
@@ -266,7 +267,7 @@ async def swagger_ui_html(req: Request) -> HTMLResponse:
                 full_openapi_url = root_path + openapi_url  # type: ignore[operator]
                 return get_swagger_ui_html(
                     openapi_url=full_openapi_url,
-                    title=self._openapi_info.title + " - Swagger UI",
+                    title=f"{self._openapi_info.title} - Swagger UI",
```
Comment (adriangb) [inline xpresso/applications.py:270]: 👍

File: xpresso/binders/_body/extractors/discriminated.py [hunk 5]
```diff
@@ -53,11 +53,15 @@ def register_parameter(self, param: inspect.Parameter) -> BodyExtractor:
...
-                    marker = param_marker
-                    break
+            marker: typing.Optional[BodyBinderMarker] = next(
+                (
+                    param_marker
+                    for param_marker in get_markers_from_parameter(sub_body_param)
+                    if isinstance(param_marker, BodyBinderMarker)
+                ),
+                None,
+            )
```
Comment (adriangb) [inline xpresso/binders/_body/extractors/discriminated.py:63]: Same as the comment above: I prefer the more verbose language neutral for loop with early exit v.s. the language specific use of `next(..., default=None)`, allocating a generator, etc.

File: xpresso/binders/_body/extractors/field.py [hunk 6]
```diff
@@ -58,11 +58,13 @@ def register_parameter(self, param: inspect.Parameter) -> BodyExtractor:
...
-                    # the outermost marker must be the field marker (us)
-                    # so the first one that isn't us is the inner marker
-                    field_marker = marker.extractor_marker
+            if (
+                isinstance(marker, BodyBinderMarker)
+                and marker.extractor_marker is not self
+            ):
+                # the outermost marker must be the field marker (us)
+                # so the first one that isn't us is the inner marker
+                field_marker = marker.extractor_marker
```
Comment (adriangb) [inline xpresso/binders/_body/extractors/field.py:67]: 👍

File: xpresso/binders/_body/extractors/form.py [hunk 7]
```diff
@@ -109,11 +109,15 @@ def register_parameter(self, param: inspect.Parameter) -> BodyExtractor:
...
-                    marker = param_marker
-                    break
+            marker: typing.Optional[BodyBinderMarker] = next(
+                (
+                    param_marker
+                    for param_marker in get_markers_from_parameter(field_param)
+                    if isinstance(param_marker, BodyBinderMarker)
+                ),
+                None,
+            )
```
Comment (adriangb) [inline xpresso/binders/_body/extractors/form.py:119]: Same as the other two comments, I think I prefer what we have now.

File: xpresso/binders/_body/extractors/form_field.py [hunk 8]
```diff
@@ -30,15 +30,15 @@ async def extract_from_form(
...
+        except InvalidSerialization as e:
             raise RequestValidationError(
                 [
                     ErrorWrapper(
                         exc=TypeError("Data is not a valid URL encoded form"),
                         loc=tuple((*loc, self.name)),
                     )
                 ]
-            )
+            ) from e
```
Comment (adriangb) [inline xpresso/binders/_body/extractors/form_field.py:41]: 👍

File: xpresso/binders/_body/extractors/form_field.py [hunk 9]
```diff
@@ -47,7 +47,7 @@ async def extract_from_form(
                         loc=tuple((*loc, self.name)),
                     )
                 ]
-            )
+            ) from exc
```
Comment (adriangb) [inline xpresso/binders/_body/extractors/form_field.py:50]: 👍

File: xpresso/binders/_body/extractors/json.py [hunk 10]
```diff
@@ -79,15 +79,15 @@ def _decode(
...
+        except Exception as e:
             raise RequestValidationError(
                 [
                     ErrorWrapper(
                         exc=TypeError("Data is not valid JSON"),
                         loc=tuple(loc),
                     )
                 ]
-            )
+            ) from e
```
Comment (adriangb) [inline xpresso/binders/_body/extractors/json.py:90]: 👍

File: xpresso/binders/_body/openapi/discriminated.py [hunk 11]
```diff
@@ -49,11 +49,7 @@ class OpenAPIContentTypeDiscriminatedMarker(OpenAPIBodyMarker):
 
     def register_parameter(self, param: inspect.Parameter) -> OpenAPIBody:
         field = model_field_from_param(param)
-        if field.required is False:
-            required = False
-        else:
-            required = True
-
+        required = field.required is not False
```
Comment (adriangb) [inline xpresso/binders/_body/openapi/discriminated.py:52]: 👍

File: xpresso/binders/_body/openapi/discriminated.py [hunk 12]
```diff
@@ -71,11 +67,15 @@ def register_parameter(self, param: inspect.Parameter) -> OpenAPIBody:
...
-                    marker = param_marker
-                    break
+            marker: typing.Optional[BodyBinderMarker] = next(
+                (
+                    param_marker
+                    for param_marker in get_markers_from_parameter(sub_body_param)
+                    if isinstance(param_marker, BodyBinderMarker)
+                ),
+                None,
+            )
```
Comment (adriangb) [inline xpresso/binders/_body/openapi/discriminated.py:77]: Same as above, let's keep what we have currently

File: xpresso/binders/_body/openapi/field.py [hunk 13]
```diff
@@ -58,11 +58,13 @@ class OpenAPIFieldMarkerBase(OpenAPIBodyMarker):
...
-                    # the outermost marker must be the field marker (us)
-                    # so the first one that isn't us is the inner marker
-                    field_marker = marker.openapi_marker
+            if (
+                isinstance(marker, BodyBinderMarker)
+                and marker.openapi_marker is not self
+            ):
+                # the outermost marker must be the field marker (us)
+                # so the first one that isn't us is the inner marker
+                field_marker = marker.openapi_marker
```
Comment (adriangb) [inline xpresso/binders/_body/openapi/field.py:67]: 👍

File: xpresso/binders/_body/openapi/file.py [hunk 14]
```diff
@@ -77,10 +77,7 @@ class OpenAPIFileMarker(OpenAPIBodyMarker):
     def register_parameter(self, param: inspect.Parameter) -> OpenAPIBody:
         field = model_field_from_param(param)
         examples = parse_examples(self.examples) if self.examples else None
-        if field.required is False:
-            required = False
-        else:
-            required = True
+        required = field.required is not False
```
Comment (adriangb) [inline xpresso/binders/_body/openapi/file.py:80]: 👍

File: xpresso/binders/_body/openapi/form.py [hunk 15]
```diff
@@ -60,8 +60,7 @@ def get_openapi_media_type(
     ) -> openapi_models.MediaType:
         encodings: typing.Dict[str, openapi_models.Encoding] = {}
         for field_name, field_openapi in self.field_openapi_providers.items():
-            encoding = field_openapi.get_encoding()
-            if encoding:
+            if encoding := field_openapi.get_encoding():
```
Comment (adriangb) [inline xpresso/binders/_body/openapi/form.py:63]: we support python 3.7 so this won't work (CI will fail momentarily)

File: xpresso/binders/_body/openapi/form.py [hunk 16]
```diff
@@ -100,21 +99,21 @@ class OpenAPIFormDataMarker(OpenAPIBodyMarker):
 
     def register_parameter(self, param: inspect.Parameter) -> OpenAPIBody:
         form_data_field = model_field_from_param(param)
-        if form_data_field.required is False:
-            required = False
-        else:
-            required = True
-
+        required = form_data_field.required is not False
```
Comment (adriangb) [inline xpresso/binders/_body/openapi/form.py:102]: 👍

File: xpresso/binders/_body/openapi/form.py [hunk 17]
```diff
@@ -100,21 +99,21 @@ class OpenAPIFormDataMarker(OpenAPIBodyMarker):
...
-                    marker = param_marker
-                    break
+            marker: typing.Optional[BodyBinderMarker] = next(
+                (
+                    param_marker
+                    for param_marker in get_markers_from_parameter(field_param)
+                    if isinstance(param_marker, BodyBinderMarker)
+                ),
+                None,
+            )
```
Comment (adriangb) [inline xpresso/binders/_body/openapi/form.py:115]: Same as above comments, let's keep what we currently have

File: xpresso/binders/_body/openapi/json.py [hunk 18]
```diff
@@ -65,10 +65,7 @@ class OpenAPIJsonMarker(OpenAPIBodyMarker):
     def register_parameter(self, param: inspect.Parameter) -> OpenAPIBody:
         examples = parse_examples(self.examples) if self.examples else None
         field = model_field_from_param(param)
-        if field.required is False:
-            required = False
-        else:
-            required = True
+        required = field.required is not False
```
Comment (adriangb) [inline xpresso/binders/_body/openapi/json.py:68]: 👍

File: xpresso/encoders/json.py [hunk 19]
```diff
@@ -74,8 +74,10 @@ def apply_custom_encoder(
         return None
 
     def __call__(
-        self, obj: Any, custom_encoder: Dict[Any, Callable[[Any], Any]] = {}
```
Comment (adriangb) [inline xpresso/encoders/json.py:77]: Good one 👍

File: xpresso/encoders/json.py [hunk 20]
```diff
@@ -137,5 +139,5 @@ def __call__(
                 data = vars(obj)
             except Exception as e:
                 errors.append(e)
-                raise ValueError(errors)
+                raise ValueError(errors) from e
```
Comment (adriangb) [inline xpresso/encoders/json.py:142]: 👍

File: xpresso/exception_handlers.py [hunk 21]
```diff
@@ -10,8 +10,7 @@
 
 async def http_exception_handler(request: Request, exc: Exception) -> JSONResponse:
     assert isinstance(exc, HTTPException)
-    headers = getattr(exc, "headers", None)
-    if headers:
+    if headers := getattr(exc, "headers", None):
```
Comment (adriangb) [inline xpresso/exception_handlers.py:13]: Can't do this because py37

File: xpresso/openapi/_builder.py [hunk 22]
```diff
@@ -45,7 +45,7 @@
         "detail": {
             "title": "Detail",
             "type": "array",
-            "items": {"$ref": REF_PREFIX + "ValidationError"},
+            "items": {"$ref": f"{REF_PREFIX}ValidationError"},
```
Comment (adriangb) [inline xpresso/openapi/_builder.py:48]: 👍, and thanks for looking though this module, I think it's very messy and probably needs several refactors to tidy up. At least it's private so we can do whatever we want in the future 😄

File: xpresso/openapi/_builder.py [hunk 23]
```diff
@@ -56,14 +56,12 @@ def get_parameters(
...
-            parameters.append(
-                dependant.openapi.get_openapi(
-                    model_name_map=model_name_map, schemas=schemas
-                )
-            )
+    parameters: List[models.ConcreteParameter] = [
+        dependant.openapi.get_openapi(model_name_map=model_name_map, schemas=schemas)
+        for dependant in deps
+        if dependant.openapi and dependant.openapi.include_in_schema
+    ]
```
Comment (adriangb) [inline xpresso/openapi/_builder.py:63]: 👍

File: xpresso/openapi/_builder.py [hunk 24]
```diff
@@ -147,16 +145,15 @@ def get_operation(
     schemas: Dict[str, Any] = {}
     route_dependant = route.dependant
     assert route_dependant is not None
-    parameters = get_parameters(
+    if parameters := get_parameters(
```
Comment (adriangb) [inline xpresso/openapi/_builder.py:148]: please refer this walrus as well

File: xpresso/openapi/_builder.py [hunk 25]
```diff
@@ -179,17 +176,18 @@ def get_operation(
     if schemas:
         components["schemas"] = {**components.get("schemas", {}), **schemas}
     http422 = str(HTTP_422_UNPROCESSABLE_ENTITY)
-    if (data.get("parameters", None) or data.get("requestBody", None)) and not any(
-        status in data["responses"] for status in (http422, "4XX", "default")
+    if ((data.get("parameters", None) or data.get("requestBody", None))) and all(
+        status not in data["responses"] for status in (http422, "4XX", "default")
```
Comment (adriangb) [inline xpresso/openapi/_builder.py:180]: 👍

File: xpresso/openapi/_builder.py [hunk 26]
```diff
@@ -179,17 +176,18 @@ def get_operation(
...
     ):
         data["responses"][http422] = {
             "description": "Validation Error",
             "content": {
                 "application/json": {
-                    "schema": {"$ref": REF_PREFIX + "HTTPValidationError"}
+                    "schema": {"$ref": f"{REF_PREFIX}HTTPValidationError"}
```
Comment (adriangb) [inline xpresso/openapi/_builder.py:186]: 👍

File: xpresso/openapi/_builder.py [hunk 27]
```diff
@@ -253,11 +252,12 @@ def filter_routes(visitor: Iterable[VisitedRoute[Any]]) -> Routes:
...
-            operations: Dict[str, Operation] = {}
-            for method, operation in path_item.operations.items():
-                if not operation.include_in_schema:
-                    continue
-                operations[method.lower()] = operation
+            operations: Dict[str, Operation] = {
+                method.lower(): operation
+                for method, operation in path_item.operations.items()
+                if operation.include_in_schema
+            }
```
Comment (adriangb) [inline xpresso/openapi/_builder.py:259]: 👍

File: xpresso/openapi/_builder.py [hunk 28]
```diff
@@ -226,17 +224,18 @@ def get_paths_items(
...
+                method.lower(): get_operation(
                     operation,
                     model_name_map=model_name_map,
                     components=components,
                     tags=tags + operation.tags,
                     response_specs={**responses, **operation.responses},
                 )
+                for method, operation in path_item.operations.items()
+                if operation.include_in_schema
+            }
```
Comment (adriangb) [inline xpresso/openapi/_builder.py:237]: 👍

File: xpresso/openapi/_responses.py [hunk 29]
```diff
@@ -38,9 +38,9 @@ def _get_response_models_from_return_hint(
                         description=description,
                         media_type=tp.media_type,
                     )
+            elif inspect.isclass(tp) and issubclass(tp, NoneType):
+                return [ResponseSpec(description=description)]
             else:
-                if inspect.isclass(tp) and issubclass(tp, NoneType):
-                    return [ResponseSpec(description=description)]
```
Comment (adriangb) [inline xpresso/openapi/_responses.py:43]: 👍  on the changes in this module, as long as it passes the tests

File: xpresso/responses.py [hunk 30]
```diff
@@ -33,11 +33,11 @@
 def get_response(request: Request) -> Response:
     try:
         return request.scope["extensions"]["xpresso"]["response"]  # type: ignore[return-type]
-    except KeyError:
+    except KeyError as e:
         raise LookupError(
             "xpresso.responses.get_response was called"
             " before the endpoint has finished executing or the endpoint raised an exception"
-        )
+        ) from e
```
Comment (adriangb) [inline xpresso/responses.py:40]: I think this is one where we're better off doing `from None`: the fact that we got a KeyError is an internal implementation detail that adds no useful information for the user, it just clutters their traceback.

**Review by adriangb — CHANGES_REQUESTED** [review]
Thank you so much for giving the codebase such an in depth look!

No worries on the size of the PR, I was a bit scared at first but after looking at the files I could tell you did a great job of sticking to style and not bundling in logic changes.

Most of the changes I'm 👍  on.

Some of them I am rejecting for performance (nit pick) or stylistic reasons. In particular, unless it is much clearer/safer to use language features (I'd say this is the case for simple list/dict comprehensions) I prefer to use language neutral control flow (I'd say this is the case for the multiline `next(..., None)` statements). This is just personal preference though, nothing wrong with your versions 😃 

Once you revert the changes that I rejected, LGTM

**Comment by yezz123** [comment]
Hello @adriangb, Thank you for the fast response, I really enjoy looking into the codebase and trying different thing with the framework, I really enjoy it in the structure while its look similar to FastAPI somehow.

I fix all the things that's not compatible with your requirements

**Comment by adriangb** [comment]
Awesome, thank you.

So my plan for this project (at least for now, not sure if it will work out long term since I haven't seen this done anywhere else) is to _release early and release often_, the goal is that _every change is a PR and every PR results in a release_. So to make this happen I'm requiring every PR to bump the project version. Could you go into `pyproject.toml` and do a patch bump (or run `poetry version patch`)? Thanks!

**Comment by yezz123** [comment]
> Awesome, thank you.
> 
> So my plan for this project (at least for now, not sure if it will work out long term since I haven't seen this done anywhere else) is to _release early and release often_, the goal is that _every change is a PR and every PR results in a release_. So to make this happen I'm requiring every PR to bump the project version. Could you go into `pyproject.toml` and do a patch bump (or run `poetry version patch`)?
> 
> There's also a couple minor listing issues 😄
> 
> Thanks!

Yes I checked the release phase, it looks good, I just push the fix, and also the fix of mypy

File: xpresso/encoders/json.py [hunk 31]
```diff
@@ -76,6 +76,8 @@ def apply_custom_encoder(
     def __call__(
         self, obj: Any, custom_encoder: Dict[Any, Callable[[Any], Any]] = {}
```
Comment (adriangb) [inline xpresso/encoders/json.py:77]:
```suggestion
        self, obj: Any, custom_encoder: Optional[Dict[Any, Callable[[Any], Any]]] = None
```

**Pull request merged**